File: src/Game/GameLoop.hs
module Game.GameLoop where

import Control.Concurrent (threadDelay)
import Engines.GetEngine
import GeniusDraw.DrawingFunctions (drawGameLoop, drawMenuGenie, drawStartGenie)
import Interface.StringsAndCommand
import Player.GetPlayerInput (getPlayerInput)

gameVersusPlayer :: [Int] -> String -> String -> IO ()
gameVersusPlayer stack namePlayer1 namePlayer2 = do
  input1 <- getPlayerInput stack (namePlayer1 ++ player1Turn)
  let stack2 = newStack stack input1
  if winned stack2
    then do
      clearT
      drawMenuGenie (namePlayer1 ++ player1Winner) wallet15Title wallet15
    else do
      input2 <- getPlayerInput stack2 (namePlayer2 ++ player2Turn)
      let stack3 = newStack stack2 input2
      if winned stack3
        then do
          clearT
          drawMenuGenie (namePlayer2 ++ player2Winner) wallet15Title wallet15
        else gameVersusPlayer stack3 namePlayer1 namePlayer2

gameVersusBot :: [Int] -> Int -> String -> IO ()
gameVersusBot stack level namePlayer = do
  let bot = getEngine level
  input <- getPlayerInput stack (namePlayer ++ player1Turn)
  let stack2 = newStack stack input
  if winned stack2
    then do
      clearT
      drawWinnerPlayer level namePlayer
    else do
      let botMove = bot stack2
      clearT
      drawGameLoop stack2 ("Hmmm! Vou tirar " ++ show (fst botMove) ++ " moedas da pilha " ++ show (snd botMove + 1))
      threadDelay 4000000

      let stack3 = newStack stack2 botMove
      if winned stack3
        then do
          clearT
          drawStartGenie gennieWinner
        else gameVersusBot stack3 level namePlayer

drawWinnerPlayer :: Int -> String -> IO ()
drawWinnerPlayer 1 namePlayer = do
  clearT
  drawMenuGenie (namePlayer ++ player1Winner) wallet10Title wallet10
drawWinnerPlayer 2 namePlayer = do
  clearT
  drawMenuGenie (namePlayer ++ player1Winner) wallet15Title wallet15
drawWinnerPlayer 3 namePlayer = do
  clearT
  drawMenuGenie (namePlayer ++ player1Winner) wallet20Title wallet20
drawWinnerPlayer x namePlayer = do
  clearT
  drawMenuGenie (namePlayer ++ player1Winner) wallet10Title wallet10

gameLoop :: [Int] -> Int -> String -> String -> IO ()
gameLoop stack level namePlayer1 "" = gameVersusBot stack level namePlayer1
gameLoop stack level namePlayer1 namePlayer2 = gameVersusPlayer stack namePlayer1 namePlayer2

newStack :: [Int] -> (Int, Int) -> [Int]
newStack (x : xs) (coins, 0) = (x - coins) : xs
newStack (x : xs) (coins, move) = x : newStack xs (coins, move - 1)

winned :: [Int] -> Bool
winned [] = True
winned (x : xs) = if x /= 0 then False else winned xs

File: prolog/src/Game/gameLoop.pl
:- style_check(-discontiguous).
:- include('../Engines/useEngine.pl').
:- include('../GeniusDraw/DrawingFunctions.pl').
:- include('../Player/GetPlayerInput.pl').
:- include('../Interface/StringsAndCommand.pl').
:- use_module(library('system')).

makePlayerMove(Stacks, Level, NamePlayer, Stacks2) :-
    player1Turn(Msg2),
    concat_atom([NamePlayer, Msg2], Msg),
    getPlayerInput(Stacks, Msg, Move),
    nth0(0, Move, Coins),
    nth0(1, Move, Stack),
    newStack(Stacks, Coins, Stack, Stacks2),
    (winned(Stacks2) -> 
    drawWinnerPlayer(Level, NamePlayer),
    halt; !).

makeBotMove(Stacks, Level, Stacks2) :-
    useEngine(Level, Stacks, Move),
    nth0(0, Move, Coins),
    nth0(1, Move, Stack),
    Stack2 is Stack + 1,
    number_string(Stack2, StackStr),
    number_string(Coins, CoinsStr),
    concat_atom(["Hmmm! Vou tirar ", CoinsStr, " moedas da pilha ", StackStr], BotMoveMsg),
    drawGameLoop(Stacks, BotMoveMsg),
    sleep(4),
    newStack(Stacks, Coins, Stack, Stacks2),
    (winned(Stacks2) -> 
    gennieWinner(WinnerMsg),
    drawStartGenie(WinnerMsg),
    halt; !).
    
gameVersusPlayer(Stacks, NamePlayer1, NamePlayer2) :-
    makePlayerMove(Stacks, 0, NamePlayer1, Stacks2),
    makePlayerMove(Stacks2, 0, NamePlayer2, Stacks3),
    gameVersusPlayer(Stacks3, NamePlayer1, NamePlayer2).

gameVersusBot(Stacks, Level, NamePlayer) :-
    makePlayerMove(Stacks, Level, NamePlayer, Stacks2),
    makeBotMove(Stacks2, Level, Stacks3),
    gameVersusBot(Stacks3, Level, NamePlayer).

drawWinnerPlayer(0, NamePlayer) :-
  player2Winner(WinnerMsg2),
  concat_atom([NamePlayer, WinnerMsg2], WinnerMsg),
  wallet15Title(WalletTitle),
  wallet15(WalletContent),
  drawMenuGenie(WinnerMsg, WalletTitle, WalletContent),
  !.

drawWinnerPlayer(1, NamePlayer) :-
  player1Winner(WinnerMsg2),
  concat_atom([NamePlayer, WinnerMsg2], WinnerMsg),
  wallet10Title(WalletTitle),
  wallet10(WalletContent),
  drawMenuGenie(WinnerMsg, WalletTitle, WalletContent),
  !.

drawWinnerPlayer(2, NamePlayer) :-
  player1Winner(WinnerMsg2),
  concat_atom([NamePlayer, WinnerMsg2], WinnerMsg),
  wallet15Title(WalletTitle),
  wallet15(WalletContent),
  drawMenuGenie(WinnerMsg, WalletTitle, WalletContent),
  !.

drawWinnerPlayer(3, NamePlayer) :-
  player1Winner(WinnerMsg2),
  concat_atom([NamePlayer, WinnerMsg2], WinnerMsg),
  wallet20Title(WalletTitle),
  wallet20(WalletContent),
  drawMenuGenie(WinnerMsg, WalletTitle, WalletContent),
  !.

gameLoop(Stacks, Level, NamePlayer1, "") :- gameVersusBot(Stacks, Level, NamePlayer1).
gameLoop(Stacks, 0, NamePlayer1, NamePlayer2) :- gameVersusPlayer(Stacks, NamePlayer1, NamePlayer2).

newStack([H | T], CoinsQuantity, 0, [NewH|T]):- NewH is H - CoinsQuantity, !.

newStack([H | T], CoinsQuantity, Index, [H | Result]):-
  Index2 is Index - 1,
  newStack(T, CoinsQuantity, Index2, Result).

winned([]):-!.
winned([0|T]) :- winned(T).
